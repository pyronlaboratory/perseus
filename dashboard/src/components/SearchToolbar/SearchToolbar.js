import React from 'react';
import PropTypes from 'prop-types';

// material imports
import { withStyles } from '@material-ui/core/styles';

import IconButton from '@material-ui/core/IconButton';
import TextField from '@material-ui/core/TextField';

// material icons
import ClearIcon from '@material-ui/icons/Clear';
import SearchIcon from '@material-ui/icons/Search';

// material datagrid
// import { GridToolbarFilterButton } from '@mui/x-data-grid';

const styles = (theme) => ({
    wrapper: {
        padding: theme.spacing(0.5, 0.5, 0),
        justifyContent: 'space-between',
        display: 'flex',
        alignItems: 'flex-start',
        flexWrap: 'wrap',
    },

    textField: {
        [theme.breakpoints.down('xs')]: {
        width: '100%',
        },
        width: '100%',
        margin: theme.spacing(2),
        '& .MuiSvgIcon-root': {
        marginRight: theme.spacing(0.5),
        },
        '& .MuiInput-underline:before': {
        borderBottom: `1px solid ${theme.palette.divider}`,
        },
    },
});

function SearchToolbar(props) {

    const { classes } = props;

    return (
        <div className={classes.wrapper} data-testid="search-toolbar" >

            {/*
            TO DO: Write a custom filter function to enable search by time 
            <div>
                <GridToolbarFilterButton />
            </div> 
            */}
            <TextField
                variant="standard"
                value={props.value}
                onChange={props.onChange}
                placeholder="Searchâ€¦"
                className={classes.textField}
                InputProps={{
                    startAdornment: <SearchIcon fontSize="small" />,
                    endAdornment: (
                        <IconButton
                            title="Clear"
                            aria-label="Clear"
                            size="small"
                            style={{ visibility: props.value ? 'visible' : 'hidden' }}
                            onClick={props.clearSearch}
                            data-testid="clear-search-button"
                        >
                            <ClearIcon fontSize="small"/>
                        </IconButton>
                    ),
                }}
            />
        </div>
    );
}

SearchToolbar.propTypes = {
    clearSearch: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    value: PropTypes.string.isRequired,
};


export default withStyles(styles)(SearchToolbar)