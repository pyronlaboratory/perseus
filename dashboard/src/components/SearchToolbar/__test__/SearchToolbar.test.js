import React, { Fragment } from 'react';
import ReactDOM from 'react-dom';

// react testing
import { render, cleanup, fireEvent } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import user from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';


// Material-UI X DataGrid doesn't allow its child components to render outside of the parent component.
// To avoid the following error while testing, import { DataGrid } from '@mui/x-data-grid'

// Error:
//   Material-UI X: Could not find the data grid context.
//   It looks like you rendered your component outside of a DataGrid or DataGridPro parent component.
//   This can also happen if you are bundling multiple versions of the data grid.

import { DataGrid } from '@mui/x-data-grid';
import SearchToolbar from './../SearchToolbar';

afterEach(cleanup);

describe('search toolbar component', () => {
    it("renders without crashing", () => {
        const div = document.createElement("div");
        const mockData = {
            rows: [
                {"id":"1", "name":"Randy Marsh"}
            ],
            columns: [
                {
                    headerName: "ID", 
                    field: "id", 
                },
                {
                    headerName: "Name",
                    field: "name"
                }
            ],
            searchText: "Randy"
        }
        ReactDOM.render(
            <Fragment>
                <DataGrid
                    components={{ Toolbar: SearchToolbar }}
                    rows={mockData.rows}
                    columns={mockData.columns}
                    componentsProps={{
                    toolbar: {
                        value: mockData.searchText,
                        onChange: (event) => requestSearch(event.target.value),
                        clearSearch: () => requestSearch(''),
                    },
                    }}
                />
            </Fragment> ,div);
    })

    it('is able to handle onChange events without failing', () => {
        const onChangeMock = jest.fn();
        const clearSearchMock = jest.fn();
        
        const mockData = {
            rows: [
                {"id":"1", "name":"Randy Marsh"}
            ],
            columns: [
                {
                    headerName: "ID", 
                    field: "id", 
                },
                {
                    headerName: "Name",
                    field: "name"
                }
            ],
            searchText: "Randy"
        }   
        render(
            <Fragment>
            <DataGrid
                components={{ Toolbar: SearchToolbar }}
                rows={mockData.rows}
                columns={mockData.columns}
                componentsProps={{
                toolbar: {
                    value: mockData.searchText,
                    onChange: (event) => onChangeMock(event.target.value),
                    clearSearch: () => requestSearch(''),
                },
                }}
            />
            </Fragment>
        );

        const input = screen.getByRole('textbox');
                
        user.type(input, mockData.searchText)
        expect(onChangeMock).toHaveBeenCalledTimes(5);

    });

    it('is able to clear queries without failing', () => {
        const onChangeMock = jest.fn();
        const clearSearchMock = jest.fn();
        
        const mockData = {
            rows: [
                {"id":"1", "name":"Randy Marsh"}
            ],
            columns: [
                {
                    headerName: "ID", 
                    field: "id", 
                },
                {
                    headerName: "Name",
                    field: "name"
                }
            ],
            searchText: "Randy"
        }   
        
        const { getByTestId } = render(
            <Fragment>
            <DataGrid
                components={{ Toolbar: SearchToolbar }}
                rows={mockData.rows}
                columns={mockData.columns}
                componentsProps={{
                toolbar: {
                    value: mockData.searchText,
                    onChange: (event) => onChangeMock(event.target.value),
                    clearSearch: () => clearSearchMock(''),
                },
                }}
            />
            </Fragment>
        );

        const input = screen.getByRole('textbox');
        
        user.type(input, mockData.searchText);
        expect(onChangeMock).toHaveBeenCalledTimes(5);
        
        const clear = getByTestId('clear-search-button');
        
        fireEvent.click(clear);
        expect(input).toHaveTextContent('');

    });

    it("matches snapshot", () => {
        const mockData = {
            rows: [
                {"id":"1", "name":"Randy Marsh"}
            ],
            columns: [
                {
                    headerName: "ID", 
                    field: "id", 
                },
                {
                    headerName: "Name",
                    field: "name"
                }
            ],
            searchText: "Randy"
        }
        const tree = renderer.create(
            <Fragment>
                <DataGrid
                    components={{ Toolbar: SearchToolbar }}
                    rows={mockData.rows}
                    columns={mockData.columns}
                    componentsProps={{
                    toolbar: {
                        value: mockData.searchText,
                        onChange: (event) => requestSearch(event.target.value),
                        clearSearch: () => requestSearch(''),
                    },
                    }}
                />
            </Fragment>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    })
});

