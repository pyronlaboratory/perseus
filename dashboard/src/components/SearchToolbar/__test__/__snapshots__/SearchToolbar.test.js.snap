// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search toolbar component matches snapshot 1`] = `null`;
